reactive-streams/src/main/scala/fs2/interop/reactivestreams/StreamSubscriber.scala

```diff
@@ -74,84 +83,124 @@ object StreamSubscriber {
-          case o         => o -> F.unit
+          case Idle(sub) =>
+            DownstreamCancellation -> (F.delay(sub.cancel) >> q.enqueue1(None.asRight))
+          case Receiving(sub) =>
+            DownstreamCancellation -> (F.delay(sub.cancel) >> q.enqueue1(None.asRight))
+          case o =>
+            o -> F.unit
         }
-        case OnDequeue(r) => {
-          case Uninitialized          => RequestBeforeSubscription(r) -> F.unit
-          case Idle(sub)              => WaitingOnUpstream(sub, r) -> F.delay(sub.request(1))
-          case err @ UpstreamError(e) => err -> r.complete(e.asLeft)
-          case UpstreamCompletion     => UpstreamCompletion -> r.complete(None.asRight)
-          case o                      => o -> r.complete((new Error(s"received request in invalid state [$o]")).asLeft)
+        case OnDequeue => {
+          case Uninitialized =>
+            RequestBeforeSubscription -> F.unit
+          case Receiving(sub) =>
+            Receiving(sub) -> F.unit
+          case err @ UpstreamError(e) =>
+            err -> (q.enqueue1(e.asLeft) >> q.enqueue1(None.asRight))
+          case Idle(sub) =>
+            Receiving(sub) -> F.delay(sub.request(maxDemand)) // request on first dequeue
+          case UpstreamCompletion =>
+            UpstreamCompletion -> q.enqueue1(None.asRight)
+          case st =>
+            st -> F.unit
         }
       }
 
-    Ref.of[F, State](Uninitialized).map { ref =>
+    Ref.of[F, State](Uninitialized) map { ref =>
       new FSM[F, A] {
-        def nextState(in: Input): F[Unit] = ref.modify(step(in)).flatten
-        def onSubscribe(s: Subscription): F[Unit] = nextState(OnSubscribe(s))
-        def onNext(a: A): F[Unit] = nextState(OnNext(a))
-        def onError(t: Throwable): F[Unit] = nextState(OnError(t))
-        def onComplete: F[Unit] = nextState(OnComplete)
-        def onFinalize: F[Unit] = nextState(OnFinalize)
-        def dequeue1: F[Either[Throwable, Option[A]]] =
-          Deferred[F, Out].flatMap(p => ref.modify(step(OnDequeue(p))).flatten >> p.get)
+        def nextState(in: Input): F[Unit] =
+          ref.modify(step(in)).flatten
+        def onSubscribe(s: Subscription): F[Unit] =
+          nextState(OnSubscribe(s))
+        def onNext(a: A): F[Unit] =
+          nextState(OnNext(a))
+        def onError(t: Throwable): F[Unit] =
+          nextState(OnError(t))
+        def onComplete: F[Unit] =
+          nextState(OnComplete)
+        def onFinalize: F[Unit] =
+          nextState(OnFinalize)
+        def dequeueChunk1: F[Chunk[Either[Throwable, Option[A]]]] =
+          for {
+            _ <- ref.modify(step(OnDequeue)).flatten
+            chunk <- q.dequeueChunk1(maxDemand)
+            _ <- ref.get flatMap {
+              case Receiving(s) => F.delay(s.request(chunk.size))
+              case Idle(s) => F.delay(s.request(chunk.size))
+              case _ => q.enqueue1(None.asRight)
```

Comment:
I think the only case we need to consider here is `Receiving`. If we've successfully dequeued something we're either in `Receiving`, in which case we request more, or we're in a terminal state in which case we don't do anything (no need to `q.enqueue1(None)` either). It isn't possible to transition from `Receiving` to `Idle` now, so we need not consider it here.